## Mean Absolute Percentage Error

**Mean Absolute Percentage Error** (MAPE) is a metric commonly used in regression tasks to evaluate the performance of a model. 

The Mean Absolute Percentage Error calculates the **average percentage difference** between the predicted values and the true values. By computing this metric, we can evaluate how well the model's predictions align with the actual values in terms of relative errors.

The formula for MAPE is as follows:

$$MAPE = \frac{1}{N}\sum_{i=1}^{N}\left(\frac{\mid y_{i} - \hat{y}_{i}\mid}{\mid y_{i}\mid}\right) \times 100$$

In this formula, $N$ represents the total number of samples, $y_{i}$ is the actual value, and $\hat{y}_{i}$ is the predicted value.

### Additional insights on Mean Absolute Percentage Error (MAPE):

1. **Application to Regression**: MAPE is commonly used as <span style="font-size: 11pt; color: green; font-weight: normal">both an evaluation metric and a loss function in regression tasks</span>.

2. **Interpretability**: <span style="font-size: 11pt; color: green; font-weight: normal">MAPE has a straightforward interpretation</span>. The value of MAPE represents the average percentage difference between the predicted and actual values. For example, an MAPE of 10 means, on average, the model's predictions deviate from the true values by 10% in terms of relative error.

3. **Percentage Error**: Unlike MAE, which measures the absolute difference, MAPE considers the relative error in terms of percentages. This property enables us to understand the magnitude of the error in relation to the actual values. MAPE is particularly useful when we want to assess the model's performance in terms of relative accuracy.

4. **Handling Zero Values**: One important consideration when using MAPE is that <span style="font-size: 11pt; color: red; font-weight: normal">MAPE is not defined when the actual value, $y_{i}$, is zero</span>. This is because the formula involves division by $y_{i}$, which would result in an undefined value. Therefore, <u>MAPE may not be suitable if the dataset contains zero values</u>. Or some other modifications to the MAPE formula, like adding eps. in denominator, must be used to avoid division by zero. 

5. **Scale Independence**: <span style="font-size: 11pt; color: green; font-weight: normal">MAPE is scale-independent</span>, meaning it does not depend on the scale of the data. <u>It allows for comparing the performance of models across different datasets or target variables</u>, as the error is expressed as a percentage.

**Since PyTorch and Scikit Learn do not provide us with the dedicated function to compute MAPE, We will create a custom function to compute Mean Absolute Percentage Error using Numpy.**

### Importing libraries

In [1]:
import numpy as np

### Generating sythetic time series data with an upward trend

In [2]:
# Simulated true and predicted values

# For SkLearn and Numpy
y_true = [100, 100, 100, 100, 100, 100, 100, 100, 100, 100]
y_pred = [80, 100, 90, 95, 105, 101, 110, 99, 87, 100]

### Computing MAPE via the custom function

In [3]:
# Create a custom function to compute sMAPE
def MAPE(y_true, y_pred):
    y_true, y_pred = np.array(y_true), np.array(y_pred)
    assert y_true.shape[0] == y_pred.shape[0], 'y_true and y_pred should have equal number of items'
    N = y_true.shape[0]
    enom = np.abs(y_true - y_pred)
    denom = np.abs(y_true)
    MAPE = np.sum((enom / denom) / N) * 100
    return round(MAPE, 2)

# Calculate SMAPE
MAPE = MAPE(y_true, y_pred)

print('Mean Absolute Percentage Error:', MAPE)

Mean Absolute Percentage Error: 6.5


MAPE of 6.5 indicates that, on average, the model's predictions deviate from the true values by approximately 6.5%.  

This means that, **for each prediction, the model's output is expected to differ from the actual value by an average of 6.5%** in terms of relative error.